### PR TITLE
[CBRD-22837] REGEXP unicode-aware enhancement

### PR DESCRIPTION
**Pull request by hgryoo**
http://jira.cubrid.org/browse/CBRD-22837

**Added locale_helper.hpp file**
Regular expression related functions utilize the helper functions in this file to make them unicode-aware. C++ `<regex>` needs wide string and std::locale class to perform locale-specific behavior. 
So helper functions provide the following.
- create and return std::locale object from the cubrid LANG_COLLATION 
- convert between Unicode (wide string) and multibyte string

Collation check for REGEXP and REGEXP_REPLACE also has been implemented.

**This PR should be merged after #1647. To review this PR, please refer the first commit of the PR which starts from 4e06348.**

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/parser/type_checking.c (2 incorrect indentation(s) found)
```
 19839       error = db_string_regexp_replace (result, args, num_args, NULL, NULL);
 19840       break;
```
#### File: src/parser/xasl_generation.c (1 incorrect indentation(s) found)
```
  6232       regu->value.funcp->tmp_obj = NULL;
```
#### File: src/query/fetch.c (2 incorrect indentation(s) found)
```
  3975 	    case F_REGEXP_REPLACE:
  4182 	case F_REGEXP_REPLACE:
```
#### File: src/query/query_executor.c (18 incorrect indentation(s) found)
```
  1533       if (regu_var->value.funcp->tmp_obj != NULL)
  1534 	{
  1535 	  switch (regu_var->value.funcp->ftype)
  1536 	    {
  1537 	    case F_REGEXP_REPLACE:
  1538 	      {
  1539 		delete regu_var->value.funcp->tmp_obj->compiled_regex;
  1540 	      }
  1541 	      break;
  1542 	    default:
  1543 	      // any member of union func_tmp_obj may have been erased
  1544 	      assert (false);
  1545 	      break;
  1546 	    }
  1547 
  1548 	  delete regu_var->value.funcp->tmp_obj;
  1549 	  regu_var->value.funcp->tmp_obj = NULL;
  1550 	}
```
#### File: src/query/query_opfunc.c (2 incorrect indentation(s) found)
```
  8521 	function_p->tmp_obj = new function_tmp_obj;
  8522 	function_p->tmp_obj->compiled_regex = new COMPILED_REGEX ();
```
#### File: src/query/stream_to_xasl.c (1 incorrect indentation(s) found)
```
  5674 
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/base/locale_helper.hpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	  case INTL_CODESET_KSC5601_EUC:
+	    conv_status = intl_euckr_to_utf8 ((const unsigned char *) in.data(), in.size(), (unsigned char **) &utf8_str_ptr,
+					      &conv_size);
+	    break;
+	  case INTL_CODESET_RAW_BYTES:
+	    intl_binary_to_utf8 ((const unsigned char *) in.data(), in.size(), (unsigned char **) &utf8_str_ptr,
+				 &conv_size);
+	    break;
+	  default:
+	    // unrecognized codeset
```
Comment (shparkcubrid) [inline src/base/locale_helper.hpp]: I think explicit error message output is required for unsupported codesets.

File: src/base/locale_helper.hpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      }
+    catch (std::exception &e)
+      {
+	// return the environment's default locale, locale name is not supported
+	assert (false);
+	return std::locale ("");
+      }
+  }
+
+  bool convert_to_wstring (std::wstring &out, const std::string &in, const LANG_COLLATION *lang_coll)
```
[the remark this answers is not in the document]
Comment (hgryoo) [reply src/base/locale_helper.hpp]: I added a comment explaining this function.

File: src/base/locale_helper.hpp (hunk 1 above)
Comment (hgryoo) [reply src/base/locale_helper.hpp]: I don't think an explicit error is required. Even If new codeset is supported in CUBRID in the future, the assertion is enough to be checked by developers.

Comment (shparkcubrid) [reply src/base/locale_helper.hpp]: i agree

File: src/query/string_opfunc.c [hunk 3]
```diff
@@ -4375,6 +4375,27 @@ db_string_rlike (const DB_VALUE * src, const DB_VALUE * pattern, const DB_VALUE
 	goto cleanup;
       }
 
+    INTL_CODESET src_codeset = db_get_string_codeset (src);
+    INTL_CODESET pattern_codeset = db_get_string_codeset (pattern);
+
+    if ((src_category != pattern_category) || (src_codeset != pattern_codeset))
+      {
```
Comment (shparkcubrid) [inline src/query/string_opfunc.c]: there is same condition above the source code. 'if (src category! = Pattern category)'. result is different

Comment (hgryoo) [reply src/query/string_opfunc.c]: i fixed it

**Pull request merged**